### ISSUE TITLE
Add support for Semigroup=>Monoid superclass

### ISSUE THREAD

**Issue by hvr** [ekmett/lens#764]
It would be useful to have this sorted out rather sooner than later so we can have a version of `lens` (even if unreleased or it only lives on the new head.hackage repo) which is GHC HEAD ready as we'd like to be able improve "early" continuous integration of GHC HEAD w/ popular Hackage packages; from the IRC channel:

```
hvr:       the problems start already w/
hvr:        instance (Contravariant f, Apply f) => Semigroup (Folding f a) where
hvr:        instance (Contravariant f, Applicative f) => Monoid (Folding f a) where
hvr:       the second instance definition fails to typecheck because it cannot infer Apply f
hvr:       so you'd have to add Apply f to the class constraints of the Monoid instance
hvr:       the same problem w/ a missing Apply constraint then propagates to other places

phadej:    having (..., Apply f, Applicative f) => Monoid (Folding f a) is IMHO good enough

edwardk:   I expect we'll have to do some surgery on lens, and that we'll lose a bit of functionality in the process
edwardk:   The decision will need to be made on each point to worsen support for third party applicatives or to worsen support for relevant traversals
edwardk:   We knew it would come up eventually

phadej:    if there aren't too many such instances, would it make sense to have two variants? one with Apply in both, and other with Applicative in both?
```

/cc @phadej @RyanGlScott @ekmett

**Comment by bgamari** [comment]
I'd like to +1 this; one of the things that came up during the Implementator's Workshop is our lack of smoke-testing against user packages between releases. Keeping packages low on the dependency tree building with `HEAD` will make testing much easier, both between and (perhaps more importantly) directly prior to releases.

**Comment by hvr** [comment]
related: #765

**Comment by RyanGlScott** [comment]
Fixed in #766.
